### PR TITLE
[Feat] Add Local Search for Solution Improvement

### PR DESCRIPTION
**Pull request by hyeok9855**
## Description

Add local search operators as a post-processing to improve a given solution.
Here, we implement [2-opt for TSP](https://github.com/henry-yeh/DeepACO/blob/main/tsp_nls/two_opt.py) and [LocalSearch operator provided by PyVRP](https://pyvrp.org/api/search.html#module-pyvrp.search.LocalSearch) for CVRP.
For other problems (e.g., PDP or scheduling), we couldn't find such a plug-and-play local search operator, and we're looking for contributions to local search for other problems!

Note that I also made some refactorings regarding type hinting.

## Motivation and Context

Local search is an essential component for an enhanced CO solver. Many research projects in the NCO field utilize local search operators, such as our recent work [GFACS](https://github.com/ai4co/gfacs), which trains NN using the solution refined by local search in an off-policy manner.

- [ ] I have raised an issue to propose this change ([required](https://github.com/ai4co/rl4co/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rl4co/envs/routing/tsp.py [hunk 1]
```diff
@@ -161,6 +174,50 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
...
+        rng = RandomNumberGenerator(kwargs.get("seed") or 0)
+        
+        action_list = actions.cpu().detach().numpy().tolist()
+        scaled_locs = scale_and_truncate_to_decimals(td["locs"].cpu().detach().numpy(), 7)
+        scaled_distmat = scale_and_truncate_to_decimals(
+            get_distance_matrix(td["locs"]).cpu().detach().numpy(), 7
+        )
+        
+        improved_action_list = []
+        for _locs, _distmat, _acts in zip(scaled_locs, scaled_distmat, action_list):
```
Comment (fedebotu) [inline rl4co/envs/routing/tsp.py:194]: If we have a batch, would it be better to call `concurrent.futures`?

File: rl4co/envs/routing/tsp.py [hunk 2]
```diff
@@ -161,6 +174,50 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
             == actions.data.sort(1)[0]
         ).all(), "Invalid tour"
 
+    @staticmethod
+    def improve_solution(td: TensorDict, actions: torch.Tensor, **kwargs) -> torch.Tensor:
```
Comment (fedebotu) [inline rl4co/envs/routing/tsp.py:178]: (as per above, we can call this `local_search`)

---

CC: @leonlan this is what we mentioned during our meeting about having an optional local search operator API in RL4CO with `pyvrp` ;  any thoughts about this?

File: tests/test_envs.py [hunk 3]
```diff
@@ -48,6 +48,10 @@ def test_routing(env_cls, batch_size=2, size=20):
     env = env_cls(num_loc=size)
     reward, td, actions = rollout(env, env.reset(batch_size=[batch_size]), random_policy)
     env.render(td, actions)
+    try:
+        env.improve_solution(td, actions)
```
Comment (fedebotu) [inline tests/test_envs.py:52]: For now this is fine, for the future I think we can move the `local_search` part in a separate test as it is optional

**Comment by fedebotu** [comment]
Great job! :rocket: 

---
Another detail I cannot review above: tt seems that the installation fails for Python=3.8. This should be because `pyvrp` is only available for Python 3.9 onwards ([reference](https://pypi.org/project/pyvrp/#:~:text=Requires%3A%20Python%20%3E%3D3.9%2C%20%3C4.0)), so maybe we can skip its for that case

File: rl4co/envs/routing/tsp.py (hunk 2 above)
Comment (leonlan) [reply rl4co/envs/routing/tsp.py:178]: We are unfortunately dropping support for the TSP-based `TwoOpt` in the next release (v0.8.0), because this operator is not very effective in VRPs. I spoke to @fedebotu in Slack and these are my suggestions:
- If you want a dedicated TwoOpt operator for TSP, you can implement this yourself. This should not be too hard (GPT-4 can give the right solution) and you can make this very performant when implemented in Cython. 
- If you want to have local search support for VRPs, the current setup can serve as a good basis. However, some details (e.g., which penalties to set in the CostEvaluator?) need to be addressed. You can always ping me to have a look when you have an implementation.

For now, I assume that you continue to plan using PyVRP v0.7.0 which has support for TwoOpt.

File: rl4co/envs/routing/tsp.py [hunk 4]
```diff
@@ -161,6 +174,50 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
...
+        """Improve the solution using local search"""
+        assert pyvrp is not None, (
+            "`pyvrp` not found. Please install pyvrp using instructions from "
+            "https://github.com/PyVRP/PyVRP or just use pip install pyvrp."
+        )
+        
+        rng = RandomNumberGenerator(kwargs.get("seed") or 0)
+        
+        action_list = actions.cpu().detach().numpy().tolist()
+        scaled_locs = scale_and_truncate_to_decimals(td["locs"].cpu().detach().numpy(), 7)
```
Comment (leonlan) [inline rl4co/envs/routing/tsp.py:188]: Support for `scale_and_truncate_to_decimals` will also be dropped in the next release. I suggest to take copy-paste the implementation.

```python
def scale_and_truncate_to_decimals(vals: np.ndarray, decimals: int = 0):
    return (vals * (10**decimals)).astype(int)
```

File: pyproject.toml [hunk 5]
```diff
@@ -59,7 +59,7 @@ dependencies = [
 graph = ["torch_geometric"]
 testing = ["pytest", "pytest-cov"]
 dev = ["black", "ruff", "pre-commit>=3.3.3"]
-routing = ["vrplib>=1.3.0"]
+routing = ["pyvrp>=0.7.0"]
```
Comment (leonlan) [inline pyproject.toml:62]: Need to pin `pyvrp==0.7.0` to guarantee that TwoOpt can be used.

File: rl4co/envs/routing/tsp.py [hunk 6]
```diff
@@ -161,6 +174,50 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
...
+            local_search.add_node_operator(TwoOpt(data=data))
+
+            # Create a pyvrp solution object from the actions
+            depot_action = _acts.index(0)
+            solution = Solution(data=data, routes=[_acts[depot_action + 1:] + _acts[:depot_action]])
+
+            # Run the local search
+            improved_solution = local_search.search(solution=solution, cost_evaluator=CostEvaluator(0, 0))
+            improved_action_list.append([0] + improved_solution.get_routes()[0].visits())
+
```
Comment (leonlan) [inline rl4co/envs/routing/tsp.py:216]: Local search looks good to me!

File: rl4co/envs/common/base.py [hunk 7]
```diff
@@ -173,11 +173,17 @@ def get_action_mask(self, td: TensorDict) -> torch.Tensor:
...
         """Function to check whether the solution is valid. Can be called by the agent to check the validity of the current state
         This is called with the full solution (i.e. all actions) at the end of the episode
         """
         raise NotImplementedError
+    
+    def improve_solution(self, td: TensorDict, actions: torch.Tensor, **kwargs) -> torch.Tensor:
+        """Function to improve the solution. Can be called by the agent to improve the current state
+        This is called with the full solution (i.e. all actions) at the end of the episode
+        """
+        raise NotImplementedError
```
Comment (fedebotu) [inline rl4co/envs/common/base.py:186]: [Minor] here we can also say for completeness

```python
raise NotImplementedError(f"Local is not implemented yet for {self.name} environment")
```

File: rl4co/envs/routing/tsp.py (hunk 1 above)
Comment (hyeok9855) [reply rl4co/envs/routing/tsp.py:194]: As pyvrp doesn't support TwoOpt in recent releases, I will replace it with the [implementation of DeepACO](https://github.com/henry-yeh/DeepACO/blob/main/tsp_nls/two_opt.py), which uses `concurrent.futures`.

File: rl4co/envs/routing/tsp.py (hunk 2 above)
Comment (fedebotu) [reply rl4co/envs/routing/tsp.py:178]: About PyVRP (CVRP, CVRPTW, new variants): we can help with that!
About PDP: from what I know, one-on-one pickup and delivery problems are not available right now, is that correct @leonlan ?

Comment (leonlan) [reply rl4co/envs/routing/tsp.py:178]: One-to-one PDP is not supported indeed.

Comment (hyeok9855) [reply rl4co/envs/routing/tsp.py]: Do you have any plan for supporting one-to-one pdp by any chance?
If so, when do you think will it be?

Comment (leonlan) [reply rl4co/envs/routing/tsp.py]: There are no plans for supporting PDP in the near future.

**Comment by hyeok9855** [comment]
### Changes
- Merged the updated main branch
- Added the reward augmentation with local search, which is one of the key components of DeepACO.

**Comment by hyeok9855** [comment]
### Changes
- Merge the main branch (c.f., #170)
- Increase the performance further (now it's very close to the performance of the original implementation of DeepACO)
- Add CVRP local search based on PyVRP

@fedebotu @leonlan 
If you have time, please review the renewed code!

File: rl4co/envs/routing/cvrp/local_search.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    seed: int = 0,
+) -> np.ndarray:
+    data = make_data(positions, demands, distances)
+    solution = make_solution(data, path)
+    ls_operator = make_search_operator(data, seed, neighbourhood_params)
+
+    improved_solution, is_feasible = perform_local_search(
+        ls_operator,
+        solution,
+        int(load_penalty * 10**4),  # * 10**4 as we scale the data by 10**4 in `make_data`
```
Comment (leonlan) [inline rl4co/envs/routing/cvrp/local_search.py:99]:
```suggestion
        int(load_penalty * 10**4),  # * 10**4 as we scale the data by 10**4 in `make_data`
```

I suggest making this value a CONSTANT variable and use it `make_data` to. Perhaps also add a comment to the CONSTANT explaining why it's needed to scale the data so that users can understand how to modify if needed.

**Comment by leonlan** [comment]
Hi @hyeok9855, I don't have time to go over the code in full detail, but I had a quick glance at the PyVRP part. It looks good to me with a small comment on the constant.

File: rl4co/envs/routing/cvrp/local_search.py (hunk 8 above)
Comment (hyeok9855) [reply rl4co/envs/routing/cvrp/local_search.py:99]: Good point, thanks!

**Comment by fedebotu** [comment]
I think we should skip the local search testing with `pyvrp` for Python < 3.9, I think that is why testing failed 

Like:
```python
@pytest.mark.skipif(sys.version_info < (3, 9))
[...]
```
Maybe in the near future we could just remove testing Python 3.8 since it's a really old version anyways

File: rl4co/models/zoo/deepaco/antsystem.py [hunk 9]
```diff
@@ -27,7 +29,10 @@ class AntSystem:
         pheromone: Initial pheromone matrix. Defaults to `torch.ones_like(log_heuristic)`.
         require_logprobs: Whether to require the log probability of actions. Defaults to False.
         use_local_search: Whether to use local_search provided by the env. Default to False.
-        local_search_params: Arguments to be passed to the local_search function.
+        use_nls: Whether to use neural-guided local search provided by the env. Default to False.
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply rl4co/models/zoo/deepaco/antsystem.py:32]: `nls` indicates the local search with neural-guided perturbation, proposed by DeepACO. See [here](https://github.com/henry-yeh/DeepACO/blob/78031e5cea30aeb9f861d90bd5a6330b8504db98/tsp_nls/aco.py#L241).

File: rl4co/envs/routing/tsp/env.py [hunk 10]
```diff
@@ -15,6 +15,7 @@
 from rl4co.utils.pylogger import get_pylogger
 
 from .generator import TSPGenerator
+from .local_search import local_search
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply rl4co/envs/routing/tsp/env.py:18]: Good, thanks!

File: rl4co/models/zoo/nargnn/encoder.py [hunk 11]
```diff
@@ -57,30 +57,34 @@ def forward(self, graph: Batch) -> Tensor:  # type: ignore
...
+            heatmap[index, edge_index[0], edge_index[1]] = edge_attr.flatten()
 
-        if self.undirected_graph:
-            heatmaps_logits = (heatmaps_logits + heatmaps_logits.transpose(1, 2)) * 0.5
+        # This is commented out, because it undo the some of the sparsification.
+        # if self.undirected_graph:
+        #     heatmap = (heatmap + heatmap.transpose(1, 2)) * 0.5
 
-        return heatmaps_logits
+        heatmap += 1e-10 if heatmap.dtype != torch.float16 else 3e-8
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply rl4co/models/zoo/nargnn/encoder.py:83]: The lowest positive value for FP32 is not 1e-10 actually. It is much smaller than that, but 1e-10 is used in DeepACO.

File: tests/test_envs.py [hunk 12]
```diff
@@ -79,6 +79,10 @@ def test_routing(env_cls, batch_size=2, size=20):
 def test_mtvrp(variant, batch_size=2, size=20):
     env = MTVRPEnv(generator_params=dict(num_loc=size, variant_preset=variant))
     reward, td, actions = rollout(env, env.reset(batch_size=[batch_size]), random_policy)
+    try:
+        env.local_search(td, actions)
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply tests/test_envs.py:83]: No, I only added LS to TSP and CVRP.
I don't remember if I added the lines. This might be due to the unexpected behavior of git merge.
I will just remove it for now!

[the remark this answers is not in the document]
Comment (hyeok9855) [reply tests/test_envs.py:83]: > We should have this available for all 16 variants with the code @leonlan made

Where can I find the code? Is that about local search?

[the remark this answers is not in the document]
Comment (fedebotu) [reply tests/test_envs.py]: It's about full solutions, but my guess is that it will work with any variant with minor modifications! [Here](https://anonymous.4open.science/r/routefinder/routefinder/baselines/pyvrp.py) is the code~

File: rl4co/envs/routing/tsp/env.py [hunk 13]
```diff
@@ -166,6 +167,19 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
             == actions.data.sort(1)[0]
         ).all(), "Invalid tour"
 
+    def generate_data(self, batch_size) -> TensorDict:
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply rl4co/envs/routing/tsp/env.py:174]: So you mean I can just remove the method?

**Comment by hyeok9855** [comment]
Unresolved comments are dealt with in the following PR.
